README.rst
======
CyCuba
======

A Cython wrapper around the Cuba multidimensional integration library.

What is Cuba?
-------------

The Cuba integration library (http://www.feynarts.de/cuba/) provides both 
Monte Carlo and deterministic rules for the evaluation of multidimensional 
integrals.

What does CyCuba provide?
-------------------------

CyCuba provides an interface between the Python interpreter and the Cuba
integration library. It allows the user to quickly and easily integrate their
Python functions, including for situations where the SciPy quadrature routines
(http://docs.scipy.org/doc/scipy/reference/integrate.html) are inappropriate due
to excessive dimensionality or non-smooth integrand functions. CyCuba is
intended to greatly simplify the process of evaluating these integrals, while
simultaneously providing transparent access to all of the different options of
the Cuba library for those users who need it.

Notably, CyCuba does NOT yet support concurrent evaluation of integrals, and
testing for some integration options remains incomplete. When you find bugs or
incomplete features that you need, please post an issue to GitHub
(https://github.com/woodscn/CyCuba), and we will try to resolve it right away.

Requirements
------------

- Cuba
- cython
- numpy
- pytest


Installation
------------

CyCuba relies on a working copy of the Cuba library. The most current version of
this library can be obtained from http://www.feynarts.de/cuba/, along with
papers which describe in detail the algorithms used in Cuba, along with the
various options and capabilities the library provides. This library and its
accompanying header files should be placed in a "normal" location for your
system, or else environment variables can be modified in order to point to their
location. For instance, if the library is placed in ``~/lib/`` and the library
header file ``cuba.h`` is placed in ``~/include/``, then one could write
(OS X or Linux): 

    export CFLAGS="-I ~/include/ $CFLAGS"
    export LDFLAGS="-L ~/lib/ $LDFLAGS"

Once this library is available, the CyCuba package can be
installed using the normal command: ``python setup.py install``. You will need
to have the ``cython``, ``numpy``, and ``pytest`` packages installed.  The
package tests may be run as ``cycuba.test()``.


Usage
-----
CyCuba provides wrappers to the four Cuba routines: Vegas, Suave, Divonne, and
Cuhre. Detailed descriptions of these routines are available in the Cuba
documentation (Ref. 1), and only the necessary Python equivalents are given
here.

Simplest possible usage:

``Vegas/Suave/Divonne/Cuhre(integrand, ranges=None)``

- ``integrand``: Python callable object. The signature is ``y = f(*args)``, where y
  is an iterable. The Python wrapper handles the necessary conversion to the
  required Cuba form. The ``userdata`` pointer is not available; if context must
  be provided to ``integrand``, then it should be defined as a callable class,
  and the context stored as state. The additional information Cuba provides,
  e.g. ``nvec``, or ``weight`` for Vegas, are not supported at this time.

- ``ranges``: Python iterable of len(2) iterables. Defines hypercubic
  integration domain, e.g. [[x0min, x0max], [x1min, x1max], ... [xnmin, xnmax]]
  If unspecified, defaults to [[0, 1], [0, 1], ...]



Common arguments:
- ``epsrel=1e-3``, ``epsabs=1e-12``: The requested relative and absolute accuracies.

- ``mineval=0``, ``maxeval=1e6``: The minimum and maximum allowed number of
  ``integrand`` evaluations.

- ``verbosity=0``: The requested verbosity level, ranging from 0 (lowest) to 3
  (highest).

- ``last_samples_only=False``: Use only the last, largest set of samples to
  compute the final integral value.

- ``do_not_smooth=False`` (Vegas and Suave only): Do not smooth the importance
  function. Recommended for non-smooth integrands.

- ``retain_state_file=False``: Do not delete the state file upon successful
  termination of the integration. Note that the use of state files is currently
  unsupported.

- ``file_grid_only=False``: Ignore integrator state, even if state file is
  present. Allows the reuse of one grid for a different integrand. Note that the
  use of state files is currently unsupported.

- ``seed=0``: Seed the random number generator. Uses Sobol quasi-random numbers
  if set to 0. If non-zero, the specific random number generator is determined
  by ``level``.

- ``level=0``: Further specify a random number generator, provided ``seed`` is
  non-zero. If ``level`` is set to zero, use Mersenne Twister. If non-zero, use
  Ranlux with a generation period ``p`` defined by ``level`` as follows:
  - 1: ``p`` = 48: Very long period; passes gap test but fails spectral test. 
  - 2: ``p`` = 97: Passes all known tests, but theoretically still defective.
  - 3: ``p`` = 223: Very small chance of observing theoretical correlations.
  - 4: ``p`` = 389: Highest possible luxury.
  - 5-23: Default to 3.
  - 24+: Specify ``p`` directly.

``def Vegas(integrand, ranges=None, **kwargs)``

Vegas-specific arguments:

  - ``nstart=1e3``: Starting number of integrand evaluations per iteration.
    
  - ``nincrease=5e2``: Increase in number of integrand evaluations per
    iteration.
    
  - ``nbatch=1e3``: Batch size for sampling.
    
  - ``gridno=0``: Slot in internal grid table. See Cuba documentation.

``def Suave(integrand, ranges=None, **kwargs)``

Suave-specific arguments:

 - ``nnew=1e3``: Number of new integrand evaluations in each subdivision.

 - ``nmin=2``: Minimum number of samples a former pass must congtribute to a
   subregion to be considered in that region's compound integration
   value. Increasing ``nmin`` may reduce jumps in the chi-squared value.

 - ``flatness=25``: The type of norm used to compute the fluctuation of a
   sample. Choose a large value for flat integrands, and a smaller value for
   volatile integrands. Very large values (> ~ 200) may cause double-precision
   overflow.

``def Divonne(integrand, ranges=None, **kwargs)``

The user is cautioned that Divonne is the most complex of the routines in Cuba,
and testing for this part of the wrapper does not provide complete coverage of
all the available options. In particular, specification of ``xgiven`` and
``peakfinder`` is untested, and may contain bugs. The developers welcome any
test routines that may be contributed to extend test coverage for Divonne (and
the other routines). 

Divonne-specific arguments (See Ref. 1 for description):

 - ``key1=47``

 - ``key2=1``

 - ``key3=1``
   
 - ``maxpass=5``

 - ``border=0``

 - ``maxchisq=10``

 - ``mindeviation=0.25``

 - ``xgiven=[]``

 - ``nextra=0``

 - ``peakfinder=None``

``def Cuhre(integrand, ranges=None, **kwargs)``

Cuhre-specific arguments:

 - ``key=1``: Select the cubature rule of degree ``key``. Available choices are
   7, 9, 11 (3-dimensions only), 13 (2-dimensions only). For other values, the
   highest available rule for the dimensionality is used.


References
----------
1. Cuba - a library for multidimensional numerical integration
(http://arxiv.org/abs/hep-ph/0404043)

2. Concurrent Cuba (http://arxiv.org/abs/1408.6373)




Cuba-4.2/ChangeLog
7 May 15:

- Fixed sign error in Cuhre's Rule.c (copy-&-paste error from
  higher-precision weights in degree-9 rule).


31 Mar 15:

- Version 4.2: Added flag for resetting the state even if a state
  file is available in Vegas (Bit 5 of 'flags').  Together with
  Bit 4 (retain state file) the grid adapted for one integrand
  can be re-used on another integrand.

- Added --with-real=10 and --with-real=16 configuration options
  for higher-precision versions of Cuba.


26 Nov 14:

- Divonne now applies Flatten to the Given list, hence either
  a list of coordinates (as before) or a list of points can be
  given.

- Tweaked MathLink code for compatibility with older Mathematica
  versions.


25 Nov 14:

- Version 4.1: Added 'ninsuf' argument to Suave.
  When computing the integral value of a subregion, samples from a 
  former pass contribute only of their number is larger than ninsuf.
  Increasing ninsuf may reduce jumps in the chi^2 value.


25 Nov 14:

- Fixed subtle bug in Cuhre which could lead to longer run times
  than in versions 1 and 2.

- Fixed miscounting of degrees of freedom in rare cases in Suave.


9 Oct 14:

- Divonne now passes the userdata argument to the peakfinder
  function, too.


27 Aug 14:

- Fixed problem due to which CUBACORESMAX was effectively ignored.

- Fixed issues with MathLink compile for Mathematica 10.


15 Aug 14:

- Version 4.0: distinguishes Accelerators (GPUs) and Cores.
  New distribution strategy (see manual), e.g. can now prescribe
  maximum number of points per core, necessary for accelerators,
  allows for load-levelling on cores.

- No. of cores/accelerators and max. points per core/accel
  can be set with environment variables
  CUBACORES (as before) and CUBACORESMAX (or API call cubacores),
  CUBAACCEL and CUBAACCELMAX (or API call cubaaccel).
  API calls generally supersede the environment.

- Can choose minimum verbosity level via environment variable
  CUBAVERBOSE = 0..3 (or API call cubaverbose).

- Can start cores/accels explicitly via cubafork/cubawait calls,
  or have the integrator keep them running at exit.


18 Apr 14:

- Fixed uninitialized variable in update of Apr 11 that could lead
  to a potential segfault in Divonne.


11 Apr 14:

- Added invocation of worker init/exit functions also for the master
  whenever the number of samples is too small to be distributed.


21 Mar 14:

- Fixed glitch in Sobol quasi-random numbers for the long long int
  versions introduced in the last update (forgotten typecast).


18 Mar 14:

- Fixed infinite loop in Sobol quasi-random numbers for maxeval
  values >~ 2^30 (32-bit versions) or 2^62 (64-bit 'll' versions).


22 Jan 14:

- Removed 'Chop' from the Mathematica sampling procedure, which would
  introduce an absolute lower scale for integrals quite close to zero.


8 Jan 14:

- Restricted detection of Qt to version 4.


18 Dec 13:

- Minor fixes in the MacOS build of partview (thanks to Hong Xu).


11 Dec 13:

- Version 3.3: Added new argument 'nvec' to let the integrand
  function have more than one point at a time, e.g. to take advantage
  of vectorization. 


5 Nov 13:

- Fixed MathLink compilation issues with MacOS 10.9/clang++.

- Fixed spurious warning in partview.


25 Sep 13:

- Fixed issue with distribution of points to cores.  Because of
  rounding up the number of points for each core it could happen
  that the last one got nothing and kept waiting indefinitely.

- Fixed bug in shared memory allocation of Divonne.

- Added Korobov tables for Divonne extending the integration
  dimension up to 257.


17 Sep 13:

- More compliance: Repaired some compiler hiccups.


7 Aug 13:

- Version 3.2: Moved all variable-size arrays to alloca and
  cleaned up other bits that would lead to compiler warnings
  on select systems.

- Cuba now compiles with clang and icc (and even C89 compilers).

- Fixed bug in chi-square determination in Cuhre introduced in 3.1.

- Fixed bug in Ranlux for luxury level > 24.


6 Jun 13:

- Fixed segfault in repeated invocation of Divonne.


4 May 13:

- Version 3.1: Added check-pointing (state file) for Suave,
  Divonne, Cuhre, too.  Implementation by B. Chokoufe.

- The Suave, Divonne, and Cuhre API calls have changed w.r.t.
  Version 3.0 ('statefile' argument added).

- Improved handling of infinities by Divonne.


28 Mar 13:

- Fixed segfault in parallel Divonne for integration rules
  (key = 7, 9, 11, 13).


25 Mar 13:

- Fixed segfault in parallel Divonne for non-Korobov random numbers.


9 Jan 13:

- Fixed bug for parallel Divonne, the 'children' fd_set was not
  properly initialized.


21 Dec 12:

- Fixed bug for parallel Divonne and ncomp > 1.
  (Workers transmitted only partial results and thus the master
  waited indefinitely for data.)


27 Nov 12:

- Fixed bug when running Divonne with ngiven != 0 and CUBACORES > 1.


6 Sep 12:

- Added functions for (de)initialization of worker processes
  (e.g. close file descriptors before exit).


3 Jul 12:

- Added fflush(NULL) before fork to avoid multiple write-out
  of buffered content at child's exit(0).


17 Apr 12:

- Added 'keep state file' flag (bit 5) to not remove the state
  file in Vegas even if the integration terminated normally.


19 Mar 12:

- Fixed two bugs in MathLink interfacing, obvious oversight from
  last release.


16 Feb 12:

- Added shared memory communication for the processes running
  in parallel.


24 Nov 11:

- Significantly rearranged Divonne, e.g. unrolled the recursion
  on regions, for better parallelizability.

- Improved parallelization through DoSample.

- Automatic detection and usage of the free cores on a machine,
  or else set CUBACORES to the number of cores one wishes to
  utilize, with negative numbers meaning "up to" depending on
  availability (e.g. -2 means "use up to two free cores").


12 Aug 11:

- Version 3.0: Added parallelization, currently 'only' through
  the DoSample routine.  Uses fork/wait so integrands do not have
  to be reentrant.


test_cycuba.py
from math import pi, sin, cos, exp, sqrt

import numpy.testing as nptest
import pytest

from cycuba import (Vegas, Suave, Divonne, Cuhre,
                    integer_bit_flags, CyCubaError, CyCubaWarning)

ndim = 3
ncomp = 1
nvec = 1
epsrel = 1e-3
epsabs = 1e-12
last = 4
seed = 0
mineval = 0
maxeval = 5e4

nstart = 1000
nincrease = 500
nbatch = 1000
gridno = 0
statefile = ""
spin = -1

nnew = 1000
nmin = 2
flatness = 25

key1 = 47
key2 = 1
key3 = 1
maxpass = 5
border = 0
maxchisq = 10
mindeviation = .25
ngiven = 0
ldxgiven = ndim
xgiven = []
nextra = 0
peakfinder = None

key = 0

verbose = 2

flag_kwargs = {'verbosity': 0, 'last_samples_only': False,
               'do_not_smooth': False, 'retain_state_file': False,
               'file_grid_only': False, 'level': 0}

common_kwargs = {'epsrel': epsrel, 'epsabs': epsabs, 'seed': seed,
                 'mineval': mineval, 'maxeval': maxeval,
                 'statefile': statefile}

vegas_results = [
    [[0.664810733790], [0.000492177172], [0.08870493338]],
    [[5.269028388903], [0.009209558720], [0.00000185723]],
    [[0.307836626649], [0.000266157600], [0.000]],
    [[0.877334733857], [0.000827895321], [0.000]],
    [[0.416407819550], [0.000280981363], [0.005]],
    [[1.201914955901], [0.001122543858], [0.000]],
    [[0.709669278182], [0.000766047026], [0.000]],
    [[0.891209461227], [0.000847259931], [0.000]],
    [[0.080172877940], [0.000060757161], [0.036]],
    [[2.396848407842], [0.003171104659], [0.000]],
    [[0.523448717427], [0.001549583946], [0.001]]
    ]

suave_results = [
    [[0.664906987056], [0.000646671743], [0.807]],
    [[5.268345737220], [0.005220179167], [1.000]],
    [[0.307754949289], [0.000289631796], [1.000]],
    [[0.877251357781], [0.000825602379], [1.000]],
    [[0.416640336680], [0.000403879196], [0.975]],
    [[1.202392100441], [0.001177754704], [1.000]],
    [[0.709620676835], [0.000708507990], [1.000]],
    [[0.891188629400], [0.000824989887], [0.136]],
    [[0.080191317183], [0.000072979494], [0.439]],
    [[2.396433599525], [0.002356525928], [1.000]],
    [[0.523384162657], [0.001087959095], [0.000]]
    ]

divonne_results = [
    [[0.664619514751], [0.000635028303], [0.000]],
    [[5.268234724759], [0.005160562827], [0.000]],
    [[0.307826360819], [0.000288507183], [0.000]],
    [[0.877397127855], [0.000846183407], [0.000]],
    [[0.416585590187], [0.000390130006], [0.000]],
    [[1.202823015224], [0.000707720777], [0.000]],
    [[0.708880815647], [0.000704254351], [0.000]],
    [[0.891110425782], [0.000352607413], [0.000]],
    [[0.080182891553], [0.000063453661], [0.000]],
    [[2.396596937191], [0.002313583295], [0.000]],
    [[0.523818896242], [0.000524028620], [0.000]]
    ]

cuhre_results = [
    [[0.664669679782], [0.000000000033], [0.000]],
    [[5.268515066720], [0.001897671739], [0.000]],
    [[0.307805926385], [0.000246474384], [0.000]],
    [[0.877461451723], [0.000805950421], [0.039]],
    [[0.416538385894], [0.000000001505], [0.000]],
    [[1.202075364130], [0.001092125490], [0.000]],
    [[0.709694848704], [0.001483691315], [0.000]],
    [[0.891212798091], [0.000000002768], [0.008]],
    [[0.080185565078], [0.000000275828], [0.000]],
    [[2.396255344838], [0.002357219909], [0.000]],
    [[0.523697555021], [0.025943077006], [0.000]]
    ]


class TestIntegrand(object):
    def __init__(self, fun):
        self.fun = fun

    def rsq(self, x, y, z):
        return x ** 2 + y ** 2 + z ** 2

    def __call__(self, x, y, z):
        if self.fun == 0:
            out = sin(x) * cos(y) * exp(z)
        elif self.fun == 1:
            out = 1 / ((x + y) ** 2 + .003) * cos(y) * exp(z)
        elif self.fun == 2:
            out = 1 / (3.75 - cos(pi * x) - cos(pi * y) - cos(pi * z))
        elif self.fun == 3:
            out = abs(self.rsq(x, y, z) - .125)
        elif self.fun == 4:
            out = exp(-self.rsq(x, y, z))
        elif self.fun == 5:
            out = 1 / (1 - x * y * z + 1e-10)
        elif self.fun == 6:
            out = sqrt(abs(x - y - z))
        elif self.fun == 7:
            out = exp(-x * y * z)
        elif self.fun == 8:
            out = x ** 2 / (cos(x + y + z + 1) + 5)
        elif self.fun == 9:
            out = 1 / sqrt(x * y * z + 1e-5) if x > .5 else sqrt(x * y * z)
        else:
            out = 1 if self.rsq(x, y, z) < 1 else 0
        return [out]


def _vegas_test_runner(ind):
    kwargs = {}
    kwargs.update(flag_kwargs)
    kwargs.update(common_kwargs)
    [integral, error, prob] = Vegas(
        TestIntegrand(ind), ranges=None,
        nstart=nstart, nincrease=nincrease,
        nbatch=nbatch, gridno=gridno,
        **kwargs)
    err_msgs = ["Integral values do not match!",
                "Error values do not match!",
                "P-values do not match!"]
    for item in zip(  # Not comparing prob values.
            [integral, error], vegas_results[ind], err_msgs):
        nptest.assert_allclose(
            item[0], item[1], atol=1e-12, rtol=1e-3,
            err_msg="Problem #: " + str(ind) + item[2])


@pytest.mark.parametrize('ind', [0, pytest.mark.xfail(1, raises=CyCubaWarning),
                                 2, 3, 4, 5, 6, 7, 8, 9, 10])
def test_Vegas(ind):
    out = _vegas_test_runner(ind)


def _suave_test_runner(ind):
    kwargs = {}
    kwargs.update(flag_kwargs)
    kwargs['last_samples_only'] = True
    kwargs.update(common_kwargs)
    [integral, error, prob] = Suave(
        TestIntegrand(ind), ranges=None,
        nnew=nnew, nmin=nmin, flatness=flatness,
        **kwargs)
    err_msgs = ["Integral values do not match!",
                "Error values do not match!",
                "P-values do not match!"]
    for item in zip(  # Not comparing prob values.
            [integral, error], suave_results[ind], err_msgs):
        nptest.assert_allclose(
            item[0], item[1], atol=1e-12, rtol=1e-3,
            err_msg="Problem #: " + str(ind) + " " + item[2])


@pytest.mark.parametrize('ind', range(11))
def test_Suave(ind):
    out = _suave_test_runner(ind)


def _divonne_test_runner(ind):
    kwargs = {}
    kwargs.update(flag_kwargs)
    kwargs.update(common_kwargs)
    [integral, error, prob] = Divonne(
        TestIntegrand(ind), ranges=None,
        key1=key1, key2=key2, key3=key3, maxpass=maxpass, border=border,
        maxchisq=maxchisq, mindeviation=mindeviation, ngiven=ngiven,
        ldxgiven=ldxgiven, xgiven=xgiven, nextra=nextra, peakfinder=peakfinder,
        **kwargs)
    err_msgs = ["Integral values do not match!",
                "Error values do not match!",
                "P-values do not match!"]
    for item in zip(  # Not comparing prob values.
            [integral, error], divonne_results[ind], err_msgs):
        nptest.assert_allclose(
            item[0], item[1], atol=1e-12, rtol=1e-3,
            err_msg="Problem #: " + str(ind) + " " + item[2])


@pytest.mark.parametrize('ind', range(11))
def test_Divonne(ind):
    out = _divonne_test_runner(ind)


def _cuhre_test_runner(ind):
    kwargs = {}
    kwargs.update(flag_kwargs)
    kwargs['last_samples_only'] = True
    kwargs.update(common_kwargs)
    [integral, error, prob] = Cuhre(
        TestIntegrand(ind), ranges=None,
        key=key,
        **kwargs)
    err_msgs = ["Integral values do not match!",
                "Error values do not match!",
                "P-values do not match!"]
    for item in zip(  # Not comparing prob values.
            [integral, error], cuhre_results[ind], err_msgs):
        nptest.assert_allclose(
            item[0], item[1], atol=1e-12, rtol=1e-3,
            err_msg="Problem #: " + str(ind) + " " + item[2])


@pytest.mark.parametrize('ind', range(11))
def test_Cuhre(ind):
    out = _cuhre_test_runner(ind)


def test_scaling():
    kwargs = {}
    kwargs.update(flag_kwargs)
    kwargs.update(common_kwargs)
    kwargs['maxeval'] = 5e5
    def test_function(x, y):
        return [1 if 1 - x**2 - y**2 > 0 else 0]
    [integral, error, prob] = Divonne(
        test_function, ranges=[[-1, 1], [-1, 1]],
        key1=key1, key2=key2, key3=key3, maxpass=maxpass, border=border,
        maxchisq=maxchisq, mindeviation=mindeviation, ngiven=ngiven,
        ldxgiven=ldxgiven, xgiven=xgiven, nextra=nextra, peakfinder=peakfinder,
        **kwargs)
    nptest.assert_allclose(integral, pi, atol=1e-2)


cycuba.py
from functools import reduce
from inspect import signature
from os import linesep, environ
from warnings import warn, simplefilter
from operator import mul

import pytest

from _cycuba import _cuba

simplefilter('always')
# Parallelization within Cuba is not yet supported. Testing procedures are
# required in order to make certain that the Python wrapper is thread-safe,
# since the PyMultiNest project found that there were problems.
environ['CUBACORES'] = '1'

ibf_v = 0
ibf_lso = False
ibf_dns = False
ibf_rsf = False
ibf_fgo = False
ibf_l = 0

class ScaledIntegrand(object):
    def __init__(self, f, ranges):
        self.f = f
        self.ranges = ranges
        self.diffs = [range_[1] - range_[0] for range_ in ranges]
        self.jacobian = reduce(mul, self.diffs, 1)

    def __call__(self, *args):
        if len(args) != len(self.ranges):
            raise CyCubaError("different lengths of args and ranges!")
        scaled_args = [range_[0] + x * diff for (range_, diff, x) in
                       zip(self.ranges, self.diffs, args)]
        return [item * self.jacobian for item in self.f(*scaled_args)]


def integer_bit_flags(verbosity, last_samples_only, do_not_smooth_vs,
                      retain_state_file, file_grid_only_v, level):
    level_key = 3 if level > 4 and level < 24 else level
    flag_string = (
        format(level_key, 'b') +
        {True: '1', False: '0'}[file_grid_only_v] +
        {True: '1', False: '0'}[retain_state_file] +
        {True: '1', False: '0'}[do_not_smooth_vs] +
        {True: '1', False: '0'}[last_samples_only] +
        {0: '00', 1: '01', 2: '10', 3: '11'}[verbosity]
    )
    return int(flag_string, base=2)


class CyCubaError(Exception):
    pass


class CyCubaWarning(UserWarning):
    pass


class CyCubaIntegration(object):
    def __init__(self, integrand, ranges, epsrel=1e-3, epsabs=1e-12, seed=0,
                 mineval=0, maxeval=1e6, statefile=""):
        self.ndim = len(signature(integrand).parameters)
        self.ranges = ranges
        # Set things up for scaling integrands
        if self.ranges:
            self.integrand = ScaledIntegrand(integrand, ranges)
        else:
            self.integrand = integrand
        self.test_args = [0.5 for ind in range(self.ndim)]
        self.ncomp = len(self.integrand(*self.test_args))
        self.nvec = 1  # Vectorized integrands not yet supported.
        self.epsrel = epsrel
        self.epsabs = epsabs
        self.seed = seed
        self.mineval = mineval
        self.maxeval = maxeval
        self.statefile = statefile
        self.spin = 0  # Spin variable is not yet supported.
        self.empty_vegas_args = (0, 0, 0, 0)
        self.empty_suave_args = (0, 0, 0)
        self.empty_divonne_args = (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
        self.empty_cuhre_args = (0,)

    def base_args(self, flags):
        return (self.ndim, self.ncomp, self.integrand, self.nvec,
                self.epsrel, self.epsabs, flags, self.seed, self.mineval,
                self.maxeval, self.statefile, self.spin)

    def _vegas(self, flags, *args):
        vegas_args = (self.base_args(flags) +
                      args +
                      self.empty_suave_args +
                      self.empty_divonne_args +
                      self.empty_cuhre_args)
        out = _cuba('vegas', *vegas_args)
        [self.neval, self.fail, self.integral, self.error, self.prob] = out
        self._check_fail()
        return [self.integral, self.error, self.prob]

    def _suave(self, flags, *args):
        suave_args = (self.base_args(flags) +
                      self.empty_vegas_args +
                      args +
                      self.empty_divonne_args +
                      self.empty_cuhre_args)
        out = _cuba('suave', *suave_args)
        [self.neval, self.nregions, self.fail,
         self.integral, self.error, self.prob] = out
        self._check_fail()
        return [self.integral, self.error, self.prob]

    def _divonne(self, flags, *args):
        divonne_args = (self.base_args(flags) +
                        self.empty_vegas_args +
                        self.empty_suave_args +
                        args +
                        self.empty_cuhre_args)
        out = _cuba('divonne', *divonne_args)
        [self.neval, self.nregions, self.fail,
         self.integral, self.error, self.prob] = out
        self._check_fail()
        return [self.integral, self.error, self.prob]

    def _cuhre(self, flags, *args):
        cuhre_args = (self.base_args(flags) +
                      self.empty_vegas_args +
                      self.empty_suave_args +
                      self.empty_divonne_args +
                      args)
        out = _cuba('cuhre', *cuhre_args)
        [self.neval, self.nregions, self.fail,
         self.integral, self.error, self.prob] = out
        try:
            self._check_fail()
        finally:
            return [self.integral, self.error, self.prob]

    def _check_fail(self):
        if self.fail == -1:
            raise CyCubaError("Dimension out of range!")
        elif self.fail > 0:
            if self.fail == 1:
                warn("Accuracy goal not achieved!", CyCubaWarning)
            else:
                warn(
                    "Accuracy goal not achieved! " + linesep +
                    "Approximately "+str(fail)+" additional points" +
                    " required to reach desired accuracy.", CyCubaWarning)



def Vegas(integrand, ranges=None, nstart=1e3, nincrease=5e2, nbatch=1e3,
          gridno=0, verbosity=ibf_v, last_samples_only=ibf_lso,
          do_not_smooth=ibf_dns, retain_state_file=ibf_rsf,
          file_grid_only=ibf_fgo, level=ibf_l, **kwargs):
    cycuba_integration = CyCubaIntegration(integrand, ranges, **kwargs)
    flags = integer_bit_flags(
        verbosity=verbosity, last_samples_only=last_samples_only,
        do_not_smooth=do_not_smooth, retain_state_file=retain_state_file,
        file_grid_only=file_grid_only, level=level)
    out = cycuba_integration._vegas(
        flags, nstart, nincrease, nbatch, gridno)
    return out


def Suave(integrand, ranges=None, nnew=1000, nmin=2, flatness=25,
          verbosity=ibf_v, last_samples_only=ibf_lso, do_not_smooth=ibf_dns,
          retain_state_file=ibf_rsf, level=ibf_l,
          **kwargs):
    cycuba_integration = CyCubaIntegration(integrand, ranges, **kwargs)
    flags = integer_bit_flags(
        verbosity=verbosity, last_samples_only=last_samples_only,
        do_not_smooth=do_not_smooth, retain_state_file=retain_state_file,
        file_grid_only=False, level=level)
    out = cycuba_integration._suave(flags, nnew, nmin, flatness)
    return out


def Divonne(integrand, ranges=None, key1=47, key2=1, key3=1, maxpass=5,
            border=0, maxchisq=10, mindeviation=.25, xgiven=[], nextra=0,
            peakfinder=None,
            verbosity=ibf_v,
            last_samples_only=ibf_lso, retain_state_file=ibf_rsf,
            level=ibf_l,
            **kwargs):
    cycuba_integration = CyCubaIntegration(integrand, ranges, **kwargs)
    flags = integer_bit_flags(
        verbosity=verbosity, last_samples_only=last_samples_only,
        do_not_smooth=False, retain_state_file=retain_state_file,
        file_grid_only=False, level=level)
    ngiven = len(xgiven)
    ldxgiven = len(xgiven[0])
    out = cycuba_integration._divonne(
        flags, key1, key2, key3, maxpass, border, maxchisq, mindeviation,
        ngiven, ldxgiven, xgiven, nextra, peakfinder)
    return out


def Cuhre(integrand, ranges=None, key=0, verbosity=ibf_v,
          last_samples_only=ibf_lso, retain_state_file=ibf_rsf,
          **kwargs):
    if 'seed' in kwargs:
        warn("Cuhre is a deterministic routine. Ignoring value for 'seed'",
             CyCubaWarning)
    cycuba_integration = CyCubaIntegration(integrand, ranges, **kwargs)
    flags = integer_bit_flags(
        verbosity=verbosity, last_samples_only=last_samples_only,
        do_not_smooth=False, retain_state_file=retain_state_file,
        file_grid_only=False, level=0)
    out = cycuba_integration._cuhre(flags, key)
    return out

def test():
    pytest.main()
